internal/k8s/reconciler/utils.go

```diff
@@ -120,73 +120,114 @@ func routeAllowedForListenerNamespaces(ctx context.Context, gatewayNS string, al
+	fromNS := gateway.GetNamespace()
+	fromGK := metav1.GroupKind{
+		Group: gateway.GroupVersionKind().Group,
+		Kind:  gateway.GroupVersionKind().Kind,
+	}
+
+	toName := string(secretRef.Name)
+	toNS := ""
+	if secretRef.Namespace != nil {
+		toNS = string(*secretRef.Namespace)
+	}
+
+	// Kind should default to Secret if not set
+	// https://github.com/kubernetes-sigs/gateway-api/blob/ef773194892636ea8ecbb2b294daf771d4dd5009/apis/v1alpha2/object_reference_types.go#L59
+	toGK := metav1.GroupKind{Kind: "Secret"}
+	if secretRef.Group != nil {
+		toGK.Group = string(*secretRef.Group)
+	}
+	if secretRef.Kind != nil {
+		toGK.Kind = string(*secretRef.Kind)
+	}
+
+	return referenceAllowed(ctx, fromGK, fromNS, toGK, toNS, toName, c)
+}
+
 // routeAllowedForBackendRef determines whether the route is allowed
 // for the backend either by being in the same namespace or by having
 // an applicable ReferencePolicy in the same namespace as the backend.
 //
 // TODO This func is currently called once for each backendRef on a route and results
 //   in fetching ReferencePolicies more than we technically have to in some cases
 func routeAllowedForBackendRef(ctx context.Context, route Route, backendRef gw.BackendRef, c gatewayclient.Client) (bool, error) {
-	backendNamespace := ""
+	fromNS := route.GetNamespace()
+	fromGK := metav1.GroupKind{
+		Group: route.GroupVersionKind().Group,
+		Kind:  route.GroupVersionKind().Kind,
+	}
+
+	toName := string(backendRef.Name)
+	toNS := ""
 	if backendRef.Namespace != nil {
-		backendNamespace = string(*backendRef.Namespace)
+		toNS = string(*backendRef.Namespace)
 	}
 
-	// Allow if route and backend are in the same namespace
-	if backendNamespace == "" || route.GetNamespace() == backendNamespace {
+	// Kind should default to Service if not set
+	// https://github.com/kubernetes-sigs/gateway-api/blob/ef773194892636ea8ecbb2b294daf771d4dd5009/apis/v1alpha2/object_reference_types.go#L105
+	toGK := metav1.GroupKind{Kind: "Service"}
+	if backendRef.Group != nil {
+		toGK.Group = string(*backendRef.Group)
+	}
+	if backendRef.Kind != nil {
+		toGK.Kind = string(*backendRef.Kind)
+	}
+
+	return referenceAllowed(ctx, fromGK, fromNS, toGK, toNS, toName, c)
+}
+
+// referenceAllowed checks to see if a reference between resources is allowed.
+// In particular, references from one namespace to a resource in a different namespace
+// require an applicable ReferencePolicy be found in the namespace containing the resource
+// being referred to.
+//
+// For example, a Gateway in namespace "foo" may only reference a Secret in namespace "bar"
+// if a ReferencePolicy in namespace "bar" allows references from namespace "foo".
+func referenceAllowed(ctx context.Context, fromGK metav1.GroupKind, fromNamespace string, toGK metav1.GroupKind, toNamespace, toName string, c gatewayclient.Client) (bool, error) {
```

Comment:
The diff here is kinda hard to read, but this function just contains the shared logic for checking existence of an applicable ReferencePolicy that we already had for xRoute->Backend.